### PR TITLE
SpectralEntropy as measure of SpectralSimilarity 

### PR DESCRIPTION
**Pull request by Alexander-Sol**
Implementing Spectral Entropy comparison, as seen here: https://www.nature.com/articles/s41592-021-01331-z

### CODE REVIEW

**Review by trishorts — CHANGES_REQUESTED** [review]
please create a unit test where both arrays are null so that first line is tested.

Also add a check for the value. So use some dummy spectra that return an actual entropy value.

File: mzLib/Test/TestSpectralSimilarity.cs [hunk 1]
```diff
@@ -47,6 +47,8 @@ public void TestAllSpectrumSimilaritiesWithoutMzFilter()
             Assert.That(s.PearsonsCorrelation(), Is.EqualTo(-0.87).Within(0.01));
             Assert.That(s.DotProduct(), Is.EqualTo(0.17).Within(0.01));
 
+            double? test = s.SpectralEntropy();
+
```
Comment (Yuling1996) [inline mzLib/Test/TestSpectralSimilarity.cs:51]: I didn't see a real test for SpectralEntropy here.  Did I miss something? How about something like "Assert.That(test, Is.EqualTo(??).Within(??))".

Comment (Alexander-Sol) [reply mzLib/Test/TestSpectralSimilarity.cs]: Good idea Yuling, thank you. Implemented as suggested.

File: mzLib/MassSpectrometry/MzSpectra/SpectralSimilarity.cs [hunk 2]
```diff
@@ -315,6 +319,37 @@ private double[] NormalizeSpectrumSum(double[] spectrum)
             return sum;
         }
 
+        // This method should only be used with the SpectrumSum normalization method
```
Comment (trishorts) [inline mzLib/MassSpectrometry/MzSpectra/SpectralSimilarity.cs:322]: which normalization method?

File: mzLib/MassSpectrometry/MzSpectra/SpectralSimilarity.cs [hunk 3]
```diff
@@ -315,6 +319,37 @@ private double[] NormalizeSpectrumSum(double[] spectrum)
...
+        // This method should only be used when allPeaks is set to true
+        public double? SpectralEntropy()
+        {
+            if (_intensityPairs.First().Item1 == -1)
```
Comment (trishorts) [inline mzLib/MassSpectrometry/MzSpectra/SpectralSimilarity.cs:326]: can delete this first check. add comment that the test for some valid pairs is performed higher in the food chain

File: mzLib/MassSpectrometry/MzSpectra/SpectralSimilarity.cs [hunk 4]
```diff
@@ -315,6 +319,33 @@ private double[] NormalizeSpectrumSum(double[] spectrum)
             return sum;
         }
 
+        // This method should only be used with the SpectrumSum normalization method
+        // This method should only be used when allPeaks is set to true
+        public double? SpectralEntropy()
```
[the remark this answers is not in the document]
Comment (Alexander-Sol) [reply mzLib/MassSpectrometry/MzSpectra/SpectralSimilarity.cs:324]: It's possible that other normalization methods could be employed, so I don't want to throw an exception. If I could figure out how to implement a custom warning, that would be ideal.

**Review by nbollis — APPROVED** [review]

**Pull request merged**